lotman/lotman_ui.go

```diff
@@ -0,0 +1,573 @@
+	"github.com/pelicanplatform/pelican/token_scopes"
+	"github.com/pelicanplatform/pelican/utils"
+)
+
+// Given a token and a list of authorized callers, check that the token is signed by one of the authorized callers. Return
+// a pointer to the parsed token.
+func tokenSignedByAuthorizedCaller(strToken string, authorizedCallers *[]string) (bool, *jwt.Token, error) {
+	ownerFound := false
+	var tok jwt.Token
+	for _, owner := range *authorizedCallers {
+		kSet, err := server_utils.GetJWKSFromIssUrl(owner)
+		if err != nil {
+			log.Debugf("Error getting JWKS for owner %s: %v", owner, err)
+			continue
+		}
+		tok, err = jwt.Parse([]byte(strToken), jwt.WithKeySet(*kSet), jwt.WithValidate(true))
+		if err != nil {
+			log.Debugf("Token verification failed with owner %s: %v -- skipping", owner, err)
+			continue
+		}
+		ownerFound = true
+		break
+	}
+
+	if !ownerFound {
+		return false, nil, errors.New("token not signed by any of the owners of any parent lot")
+	}
+
+	return true, &tok, nil
+}
+
+// Verify that a token received is a valid token. Upon verification, we set the lot's parents/owner to the
+// appropriate values. Returns true if the token is valid, false otherwise.
+func VerifyNewLotToken(lot *Lot, strToken string) (bool, error) {
+	tokenApproved := false
+
+	// Get the path associated with the lot (right now we assume/enforce a single path) and try
+	// to deduce a namespace prefix from that. We'll use that namespace prefix's issuer as the
+	// owner of the lot. If there is no associated isuer, we assign ownership to the federation
+
+	path := ((lot.Paths)[0]).Path
+	log.Debugf("Attempting to add lot for path: %s", path)
+	errMsg := make([]byte, 2048)
+	lots := unsafe.Pointer(nil)
+	// Pass a pointer to the first element of the slice to the C++ function.
+	ret := LotmanGetLotsFromDir(path, false, &lots, &errMsg)
+	if ret != 0 {
+		trimBuf(&errMsg)
+		return false, errors.Errorf("Error getting lot JSON: %s", string(errMsg))
+	}
+
+	goLots := cArrToGoArr(&lots)
+
+	// Check if goOut[0] is "default". Lotman should ALWAYS put something in this array. If it's empty, we have a problem.
+	if goLots[0] == "default" {
+		// We'll assign to the root lot. However, be careful here because we make an assumption -- under the hood,
+		// Lotman will return "default" if there's no other lot, or if the actual path is in some way assigned to
+		// the default lot. When we assign to parent to root here, we're assuming the former. This implies we should stay
+		// away from assigning paths to the default lot.
+		lot.Parents = []string{"root"}
+	} else {
+		// We found a different logical parent
+		lot.Parents = goLots
+	}
+
+	var tok jwt.Token
+	if len(lot.Parents) == 0 && lot.Parents[0] == "root" {
+		// We check that the token is signed by the federation
+		// First check for discovery URL and then for director URL, both of which should host the federation's pubkey
+		issuerUrl := param.Federation_DiscoveryUrl.GetString()
```

Comment:
See my earlier comment about making this issuerURL retrieval and utility function.